File: apps/web/src/types/progress.ts
import type {
  CompletionHistory,
  DayHistory,
} from "./chore";


export type ProgressDayApiResponse = {
  progress_date: string;
  total_count: number;
  completed_count: number;
  all_completed: boolean;
  updated_at: string;
};


export type ProgressHistoryApiResponse = {
  history:
    ProgressDayApiResponse[];

  current_streak: number;
  longest_streak: number;
  perfect_days: number;
  average_completion: number;
  recorded_days: number;
};


export type ProgressHistory = {
  history: CompletionHistory;
  currentStreak: number;
  longestStreak: number;
  perfectDays: number;
  averageCompletion: number;
  recordedDays: number;
};


export type {
  CompletionHistory,
  DayHistory,
};

File: apps/api/app/schemas/progress.py
from datetime import date, datetime

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    model_validator,
)


class ProgressSnapshotUpsert(
    BaseModel,
):
    progress_date: date

    total_count: int = Field(
        ge=0,
        le=100_000,
    )

    completed_count: int = Field(
        ge=0,
        le=100_000,
    )

    @model_validator(
        mode="after",
    )
    def validate_counts(
        self,
    ) -> "ProgressSnapshotUpsert":
        if (
            self.completed_count
            > self.total_count
        ):
            raise ValueError(
                "Completed count cannot exceed "
                "total count.",
            )

        return self


class ProgressImportRequest(
    BaseModel,
):
    snapshots: list[
        ProgressSnapshotUpsert
    ] = Field(
        max_length=1_000,
    )


class ProgressDayResponse(
    BaseModel,
):
    model_config = ConfigDict(
        from_attributes=True,
    )

    progress_date: date
    total_count: int
    completed_count: int
    all_completed: bool
    updated_at: datetime


class ProgressHistoryResponse(
    BaseModel,
):
    history: list[
        ProgressDayResponse
    ]

    current_streak: int
    longest_streak: int
    perfect_days: int
    average_completion: int
    recorded_days: int
